proposals/text/3875-table-span.md

```diff
@@ -0,0 +1,228 @@
+
+## Code changes
+- Addition of `TableSpan` dataclass to https://github.com/deepset-ai/haystack/blob/main/haystack/schema.py
+```python
+@dataclass
+class TableSpan:
+    col: int
+    row: int
+    """
+    Defining a table cell via the column and row index.
+
+    :param col: Column index of the cell
+    :param row: Row index of the cell
+    """
+```
+
+- Updating code (e.g. schema objects, classes, functions) that use `Span` to also support `TableSpan` where appropriate.
+This includes:
+- Updating the `Answer` dataclass to support `TableSpan` as a valid type for `offsets_in_document` and `offsets_in_context`
+```python
+@dataclass
+ class Answer:
+     answer: str
+     type: Literal["generative", "extractive", "other"] = "extractive"
+     score: Optional[float] = None
+     context: Optional[Union[str, pd.DataFrame]] = None
+     offsets_in_document: Optional[List[Span], List[TableSpan]] = None
+     offsets_in_context: Optional[List[Span], List[TableSpan]] = None
+     document_id: Optional[str] = None
+     meta: Optional[Dict[str, Any]] = None
+```
+- Updating any functions that accept table answers as input to use the new `col` and `row` variables instead of `start` and `end` variables.
+This type of check for table answers is most likely already done by checking if the `context` is of type `pd.DataFrame`.
+- `TableReader` and `RCIReader` to return `TableSpan` objects instead of `Span`
+
+## Edge Case/Bug
+Internally, Haystack stores a table as a pandas DataFrame in the `Answer` dataclass, which does not treat the column
+labels as the first row in the table.
+However, in Haystack's rest-api the table is converted into a list of lists format where the column labels are
+stored as the first row, which can be seen [here](https://github.com/deepset-ai/haystack/pull/3872), which is consistent
+with the `Document.to_dict()` method seen [here](https://github.com/deepset-ai/haystack/blob/6af4f14fe0d375a1ae0ced18930a9239401231c7/haystack/schema.py#L164-L165).
+
+This means that the current `Span` and (new) `TableSpan` dataclass point to the wrong location when the table is
+converted to a list of lists.
+
+For example, the following code
+```python
+import pandas as pd
+from haystack import Document
+
+data = {
+    "actors": ["brad pitt", "leonardo di caprio", "george clooney"],
+    "age": ["58", "47", "60"],
+    "number of movies": ["87", "53", "69"],
+    "date of birth": ["18 december 1963", "11 november 1974", "6 may 1961"],
+}
+table_doc = Document(content=pd.DataFrame(data), content_type="table")
+span = (0, 0)
+print(table_doc.content.iloc[span])  # prints "brad pitt"
+
+dict_table_doc = table_doc.to_dict()
+print(dict_table_doc["content"][span[0]][span[1]])  # prints "actors"
+```
+
+I see a few ways to address this:
+- One way to address this would be to update the `Span` (or `TableSpan`) to be updated when the table is converted into a list of lists.
+- Another way would be to only store the table internally as a list of lists.
```

Comment:
How about removing the "list-of-list"s conversions? I'd imagine we can just always use Pandas DF also in the REST API, or are there cases where this can't be done?

Reply:
I don’t think this can be done because I’m not sure that pandas dataframes are json serializable. Or if they are they probably are converted into a list of lists or a list of dicts. So we make it explicit in our rest api to choose the list of lists format. 

Similarly when using a the `to_dict` method of Documents we need to choose how we save the pandas dataframe as a json serializable object.

Reply:
@julian-risch Could I get your opinion on how you think we should address this bug?

Reply:
The second and the third option sound okay to me: 

> - Only store the table internally as a list of lists. We can't only use pandas dataframes since we need to be able json serialize the table to be used in the rest-api and to be compatible with our `to_dict` and `from_dict` methods.
> - Expand `TableCell` to have additional member variables like `pandas_row`, `pandas_col` in addition to the normal `row` and `col`.

Updating `row` and `col` is too confusing in my opinion. One more idea might be to split the table into a table header and a table body.

Reply:
list of lists format is our best option in my opinion.

Reply:
Regarding serialization of the dataframes: there is [df.to_json](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_json.html) and [df.read_json ](https://pandas.pydata.org/docs/reference/api/pandas.read_json.html) for that purpose but it might slightly change the data.

Reply:
We could switch to using `df.to_json` and `df.read_json`. If we did this we'd need to probably settle on one of their output formats, which is controlled by the `orient` variable. Unfortunately, as you said this would change the format in which we export the table when using our `to_dict` and `from_dict` methods. And this would cause a breaking change in our API.

Reply:
> list of lists format is our best option in my opinion.

I agree with this. I also think updating `row` and `col` would be confusing. I think this should not be too difficult to implement. However, this would result in something like a breaking change because users would no longer be able to access the table as a pandas dataframe from within a Haystack Document, but I'm not sure if this will really impact too many people.

Reply:
> Only store the table internally as a list of lists

I didn't understand the full implication of this. Is it that any time `Answer` is created for tables, we store the context as list of lists? So invocation of the TableReader in code (not through the REST API) would also return `answer.context: List[List[str]]`?

Reply:
Yes, that is correct @bglearning. The idea would be to always store the `answer.context` as a `List[List[str]]`.

Reply:
Oh, got it. That is quite a bit change though. To confirm, it's not part of the proposal, right? Only that if we wanted to handle that edge-case it'd be one of the (possibly best) way to go. For the scope of the implementation of this proposal, the edge-case would be present and the user would have to be aware of/informed about the (row, col) not including the first list (header)?

Reply:
If possible I would like to include this in the proposal since we already discussing it and I would say it is a fairly large bug when it would come to users using TableQA through the rest API.

Reply:
Oh okay. Then perhaps the Basic Example section would need to updated [1]. And maybe needs one more look from core if we are to change `Document.content` itself and not only `Answer.context` as again, seems a big change to a core class/schema. Plus might need a longer deprecation cycle 😄  @julian-risch 

Though I guess at least within haystack, only the TableQA parts rely on content being DataFrame which as you say can be updated to have a to_df conversion before the model call.

[1] 
To make sure I understand. The approach would be that, given:

| |Actor | Age| No. Movies|
|---|------|-----|-----|
|0|Brad Pitt | 58|87|
|1|Leo | 47|53|

If say the answer is "58"

```python
prediction = reader.predict(...)
answer = prediction["answers"][0]

# New feature
# answer.context -> [["Actor", "Age", "No. Movies"], ["0", "Brad Pitt",...], [...]]
# answer.offsets_in_context[0] -> (row=1, col=1)
print(answer.context[answer.offsets_in_context[0].row][answer.offsets_in_context[0].col])
```

So essentially the offset's row would start with the first list (i.e. column) which would be handled by having `_calculate_answer_offsets` add 1 to the row coordinate returned by the model (which doesn't cover the header).

Reply:
Yes @bglearning that is exactly right! If @julian-risch agrees this would be all right to have in this proposal as well, then I'll add your example to the Basic Example section!

Reply:
Yes, makes sense to me. 👍

Reply:
Thinking out loud: if we are revisiting the core schema, wonder if it makes sense to think of what's a good easily serializable representation for a table. The `List[List]` is the simplest but maybe we could tradeoff some simplicity to reduce ambiguity. 

E.g. `{'columns': List[str], 'data': List[List]}`. Wouldn't have to (remember to) treat the first row in a special way + easier to do things like representing a "header"-less table (don't have to insert a numbered row at the start). And the conversion to DataFrame is more natural with `pd.DataFrame(**the_dict)`

Or one of the dataframe dict formats/orientation.

Reply:
I think this discussion about changing the core schema to a different format like `{'columns': List[str], 'data': List[List]}` is worth a separate proposal/discussion since that will require changes to the Rest-API and our `Document.to_dict` and `Document.from_dict` methods to the new format. 

Part of the reason I think `List[List]` is easiest for now, is because we are already using it in our rest-API and is the current export format in `to_dict`. So making this change will fix an existing bug with minimal breaking changes. 

And to address some of your specific points:
> Wouldn't have to (remember to) treat the first row in a special way 

That is true, but we only have to treat it specially when creating the `Answer` object which only occurs in a few places and should only affect developers adding new TableReaders. End-users will never need to worry about this. 

> + easier to do things like representing a "header"-less table (don't have to insert a numbered row at the start).

 If we use the export methods of pandas they would already insert a numbered row for the column headers for a "header"-less table (e.g. `{'columns': [0, 1], 'data': ["apple", "pear"]}`. 

Overall though I agree this is worth further discussion in a new proposal. And even if we do decide on a new format like the one you suggested from pandas, we can easily update how the offsets in answers are calculated so they point to the correct table cell. End-users would likely not notice that the offsets have changed since they still point to the correct answer cell.

Reply:
Makes sense 👍 

I was thinking along the lines of: if we are changing the `content` field itself, we might as well think about more generally what's a good representation for table data. 

But it's a very good point that `List[List]` does bring about far fewer breaking public API changes, (I think) namely only in the direct access of `Document.content` and `Answer.context`. So any diversion from that needs further justification over the current one for `pd.DataFrame` -> `List[List]` i.e. easier serializability and consistency.

Reply:
@bglearning I opened an issue for that future discussion [here](https://github.com/deepset-ai/haystack/issues/3999). I copied over some of our discussion.